docs/JLBP-0022.md

```diff
@@ -0,0 +1,56 @@
+---
+jlbp:
+  id: JLBP-22
+permalink: /JLBP-22
+---
+
+# Declare all dependencies
+
+If your code references a class—for example, by invoking a method in that class—declare a dependency that includes that class in your pom.xml, build.gradle,
+or equivalent. In Bazel this practice is called "strict deps".
+
+Code should not call methods, reference fields, or instantiate classes from _indirect_ dependencies. These are dependencies of the declared dependencies. Projects that rely on indirect dependencies have an annoying habit of breaking in unexpected ways when direct dependencies are upgraded. 
+
+For example, your project might declare a dependency on the 
+Google HTTP Java Client which
```

Comment:
```suggestion
Google HTTP Java Client, which
```

Reply:
I think this is a restrictive clause, but it's arguable.

Reply:
If it were restrictive, that would mean that there are several Google HTTP Java Clients, only one of which depends on Apache HTTP Components. Is that the case?

Reply:
That's one way of identifying restrictive subordinate clauses, but a little more generally, "A restrictive adjective clause, on the other hand, is essential to a sentence and should not be set off by commas" <https://www.thoughtco.com/restrictive-and-nonrestrictive-adjective-clauses-1689689>

I think this clause is indeed essential to the sentence.

A better reference than that will have to wait until I get back to the office to retrieve my Chicago Manual of Style. :-)

Reply:
The content of the clause is important to the meaning of the paragraph, but not essential to the sentence.

"For example, your project might declare a dependency on the Google HTTP Java Client."

That doesn't change the reader's understanding of what you mean by "the Google HTTP Java Client", although it leaves out an important detail about why you're talking about it.

Compare to your sentence above: "Projects that rely on indirect dependencies have an annoying habit of breaking in unexpected ways when direct dependencies are upgraded." If you got rid of the restrictive clause and changed it to "Projects have an annoying habit of breaking in unexpected ways when direct dependencies are upgraded", the reader wouldn't know which projects you were talking about.

Anyway, it's your document.

Reply:
The point is "essential to a sentence," which clause this is. It is not whether it is essential to the meaning of the modified noun or noun phrase.